File: src/main/scala/avalon/group/GroupMessageResponder.scala
package avalon.group

import java.util.regex.Pattern

import avalon.util.{BasicResponder, GroupConfig, GroupMessage}
import org.apache.commons.lang3.builder.{EqualsBuilder, HashCodeBuilder}
import org.jetbrains.annotations.NotNull

/**
	* Created by Eldath on 2017/1/28 0028.
	*
	* @author Eldath
	*/
trait GroupMessageResponder extends BasicResponder {

	def doPost(@NotNull message: GroupMessage, @NotNull groupConfig: GroupConfig): Unit

	override def getHelpMessage: String

	override def getKeyWordRegex: Pattern

	def permissionIdentifier: Array[String] = null

	override def equals(obj: scala.Any): Boolean = {
		if (obj == null || (!obj.isInstanceOf[GroupMessageResponder])) return false
		val that = obj.asInstanceOf[GroupMessageResponder]
		new EqualsBuilder()
			.append(getKeyWordRegex, that.getKeyWordRegex)
			.append(permissionIdentifier, that.permissionIdentifier)
			.isEquals
	}

	override def hashCode(): Int =
		new HashCodeBuilder(17, 37)
			.append(getHelpMessage)
			.append(getKeyWordRegex)
			.append(permissionIdentifier)
			.hashCode()

	def instance: GroupMessageResponder
}

File: src/main/kotlin/avalon/group/GroupMessageResponder.kt
package avalon.group

import avalon.util.GroupConfig
import avalon.util.GroupMessage
import org.apache.commons.lang3.builder.EqualsBuilder
import org.apache.commons.lang3.builder.HashCodeBuilder
import java.util.regex.Pattern

abstract class GroupMessageResponder {
	abstract fun doPost(message: GroupMessage, groupConfig: GroupConfig)

	abstract fun instance(): GroupMessageResponder?

	abstract fun responderInfo(): ResponderInfo

	override fun equals(other: Any?): Boolean {
		if (this === other || other === null) return true
		if (javaClass != other.javaClass) return false
		other as GroupMessageResponder
		val info = other.responderInfo()
		return EqualsBuilder()
				.append(info.helpMessage, responderInfo().helpMessage)
				.append(info.keyWordRegex, responderInfo().keyWordRegex).isEquals
	}

	override fun hashCode(): Int =
			HashCodeBuilder(17, 37)
					.append(responderInfo().helpMessage)
					.append(responderInfo().keyWordRegex.toString()).toHashCode()
}

enum class ResponderPermission {
	ALL, ADMIN, OWNER
}

class ResponderInfo(val helpMessage: Pair<String, String>,
                    val keyWordRegex: Pattern,
                    val configIdentifier: Array<String>? = null,
                    val manageable: Boolean = true,
                    val permission: ResponderPermission = ResponderPermission.ALL)